java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestRetryWithExpiredToken.java

```diff
@@ -0,0 +1,241 @@
+          return;
+        }
+        final Schema pojoSchema = new Schema(ImmutableList.of(Field.nullable("a",
+                Types.MinorType.BIGINT.getType())));
+        try (VectorSchemaRoot root = VectorSchemaRoot.create(pojoSchema, allocator)) {
+          listener.start(root);
+          root.allocateNew();
+          root.setRowCount(4095);
+          listener.putNext();
+          listener.completed();
+        }
+      }
+    };
+  }
+
+  private void startServerAndClient() throws IOException {
+    final FlightProducer flightProducer = getFlightProducer();
+    this.server = FlightTestUtil.getStartedServer((location) -> FlightServer
+            .builder(allocator, location, flightProducer)
+            .headerAuthenticator(new GeneratedTestBearerTokenAuthenticator(
+                    new BasicCallHeaderAuthenticator(this::validate)))
+            .build());
+
+    this.client = FlightClient.builder(allocator, server.getLocation())
+            .build();
+  }
+
+  @After
+  public void shutdown() throws Exception {
+    AutoCloseables.close(client, client2, server, allocator);
+    client = null;
+    client2 = null;
+    server = null;
+    allocator = null;
+  }
+
+  private CallHeaderAuthenticator.AuthResult validate(String username, String password) {
+    if (Strings.isNullOrEmpty(username)) {
+      throw CallStatus.UNAUTHENTICATED.withDescription("Credentials not supplied.").toRuntimeException();
+    }
+    final String identity;
+    if (USERNAME_1.equals(username) && PASSWORD_1.equals(password)) {
+      identity = USERNAME_1;
+    } else {
+      throw CallStatus.UNAUTHENTICATED.withDescription("Username or password is invalid.").toRuntimeException();
+    }
+    return () -> identity;
+  }
+
+  @Test
+  public void testListFlightsWithRetry() {
+    client.authenticateBasicToken(USERNAME_1, PASSWORD_1);
+    Iterable<FlightInfo> flights = client.listFlights(Criteria.ALL);
+    int count = 0;
+    for (FlightInfo flight : flights) {
+      count += 1;
+      Assert.assertArrayEquals(flight.getDescriptor().getCommand(), "flight1".getBytes(Charsets.UTF_8));
+    }
+    Assert.assertEquals(1, count);
+  }
+
+  @Test
+  public void testListActionsWithRetry() {
+    client.authenticateBasicToken(USERNAME_1, PASSWORD_1);
+    Assert.assertFalse(ImmutableList.copyOf(client
+            .listActions())
+            .isEmpty());
+    Assert.assertFalse(ImmutableList.copyOf(client
+            .listActions())
+            .isEmpty());
+  }
+
+  @Test
+  public void testDoActionWithRetry() {
+    client.authenticateBasicToken(USERNAME_1, PASSWORD_1);
+    Assert.assertFalse(ImmutableList.copyOf(client
+            .doAction(new Action("hello")))
+            .isEmpty());
+    Assert.assertFalse(ImmutableList.copyOf(client
+            .doAction(new Action("world")))
+            .isEmpty());
+  }
+
+
+  /**
+   * Generates and caches bearer tokens from user credentials.
```

Comment:
Can we make it clear here that this class generates bearer tokens with a sequentially increasing ID, and it considers token #1 to be 'expired'? It was a little unclear how the retry was getting tested at first.